src/Lean/Elab/Tactic/BVDecide/Frontend/Normalize/AC.lean

```diff
@@ -1,39 +1,362 @@
 /-
-Copyright (c) 2025 Lean FRO, LLC. All rights reserved.
+Copyright (c) 2024 Lean FRO, LLC. All rights reserved.
 Released under Apache 2.0 license as described in the file LICENSE.
-Authors: Henrik Böving
+Authors: Alex Keizer, Siddharth Bhat
 -/
 prelude
-import Lean.Elab.Tactic.BVDecide.Frontend.Normalize.Basic
 import Lean.Meta.Tactic.AC.Main
+import Lean.Elab.Tactic.BVDecide.Frontend.Normalize.Basic
+import Init.Grind.Lemmas
+
+namespace Lean.Elab.Tactic.BVDecide
+namespace Frontend.Normalize
+
+open Lean Meta
+
+/-! ### Types -/
+
+abbrev VarIndex := Nat
+
+@[match_pattern]
+def mkBitVec (w : Expr) := mkApp (.const ``BitVec []) w
+
+/-- Bitvector operations that we perform AC canonicalization on. -/
+inductive Op
+| mul (w : Expr) (inst : Expr)
+deriving BEq, Repr
+
+namespace Op
+
+/-- Given an expression of an (unapplied) operation, return the decoded `Op`.
+
+Return `none` if the expression is not a recognized operation. -/
+def ofExpr? (e : Expr) : Option Op :=
+  match_expr e with
+  | HMul.hMul bv _bv _bv inst =>
+    match bv with
+    | mkBitVec w => some (.mul w inst)
```

Comment:
Why do we need to have a special match pattern construct here instead of:
```
match_expr bv with
| BitVec w => some (.mul w inst)
```

Note that you could even use two `let_expr` with ` | return none` to implement this.

Reply:
I kept the top-level as a match, as it will have more alternatives in the future, but the second level I replaced with a let_expr as suggested